pkg/controller/moveengine/controller.go

```diff
@@ -126,48 +134,211 @@ func (r *ReconcileMoveEngine) Reconcile(request reconcile.Request) (reconcile.Re
 		return reconcile.Result{}, err
 	}
 
-	if instance.Status.Status == "" {
-		err := r.ddm.Init(instance.Spec.PluginProvider, *instance)
+	// TODO
+	me, err := engine.NewMoveEngineAction(r.log, r.client, r.discoveryHelper, instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if EnableResources {
+		err = me.ParseResourceEngine(instance)
+		if err != nil {
+			r.log.Error(err, "Failed to parse resources")
+		}
+	}
+
+	switch instance.Status.Status {
+	case "":
+		// We haven't invoked the INIT API yet. Let's initialize the MoveEngine.
+		err := r.initializeMoveEngine(me)
 		if err != nil {
-			r.log.Error(err, "Failed to initialize plugin")
 			return reconcile.Result{}, err
 		}
+		// MoveEngine has been initialized. Rest of the process should execute on next reconcilation.
+		return reconcile.Result{Requeue: true}, nil
+
+	case v1alpha1.MoveEngineInitialized:
+		requeue, requeueAfter, err := r.ensureMoveEngineReady(me)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if requeue {
+			return reconcile.Result{Requeue: requeue, RequeueAfter: requeueAfter}, nil
+		}
+	case v1alpha1.MoveEngineInitializationFailed:
+		// MoveEngine has failed to be initialized. There is no point of processing further.
+		r.log.Info("Skipping processing MoveEngine: %s/%s. Reason: MoveEngine has failed to be initialized.", instance.Namespace, instance.Name)
+		return reconcile.Result{}, nil
+	case v1alpha1.MoveEngineReady:
+		// MoveEngine is ready. Now, we can start rest of the sync process.
+
```

Comment:
It will be better if we move below code to a separate function.

Reply:
done